src/server/api/user.ts

```diff
@@ -1,192 +1,144 @@
-    res.badRequest(
-      jsonMessage(`state parameter must be one of {${ACTIVE}, ${INACTIVE}}.`),
-    )
+  if (req.files) {
+    req.body.files = req.files
   }
-
   next()
 }
 
 /**
- * Make sure all parameters needed for the pre-signed url request are present.
+ * Endpoint for a user to create a short URL. A short URL can either point to
+ * a long URL, or a file provided by the user.
  *
- * @param {string} fileType - File type of the file that is being uploaded.
- * This must be declared here so that subsequent PUT requests that use the
- * `Content-Type` header will pass header checks.
- * @param {string} key - Name of the entry to be inserted to the S3 bucket. Ensure
- * that the key does not collide with other files before declaring it here. Otherwise,
- * it will cause an existing file of the same key to be overridden.
+ * In order to create a file URL, provide the `isFile` parameter set to `true`, and supply
+ * the binary in a multipart/form-data under the field titled `file`. The longUrl
+ * parameter in the request body must point to the s3 bucket with the correct key. For
+ * example, a file upload request for the shortUrl named `test` should have `longUrl` set
+ * to `https://file[-staging].go.gov.sg/test`.
  */
-function validatePresignedUrlRequest(
-  req: Express.Request,
-  res: Express.Response,
-  next: Express.NextFunction,
-) {
-  if (!req.body.fileType || !req.body.key) {
-    return res.badRequest(
-      jsonMessage('Some or all required arguments are missing: fileType, key.'),
-    )
-  }
-  return next()
-}
-
-/**
- * Endpoint for a user to create a short URL.
- */
-router.post('/url', validateUrls, async (req, res) => {
-  const { isFile, userId, longUrl, shortUrl } = req.body
+router.post(
+  '/url',
+  fileUploadMiddleware,
+  preprocessPotentialIncomingFile,
+  validator.body(urlSchema),
+  async (req, res) => {
+    const { userId, longUrl, shortUrl } = req.body
+    const file = req.files?.file
+
+    try {
+      const user = await User.findByPk(userId)
+
+      if (!user) {
+        res.notFound(jsonMessage('User not found'))
+        return
+      }
 
-  try {
-    const user = await User.findByPk(userId)
+      const existsShortUrl = await Url.findOne({ where: { shortUrl } })
+      if (existsShortUrl) {
+        res.badRequest(jsonMessage(`Short link "${shortUrl}" already exists`))
+        return
+      }
 
-    if (!user) {
-      res.notFound(jsonMessage('User not found'))
-      return
-    }
+      // Success
+      const { buildFileLongUrl, uploadFileToS3 } = container.get<S3Interface>(
+        DependencyIds.s3,
+      )
+      const result = await transaction(async (t) => {
+        const url = Url.create(
+          {
+            userId: user.id,
+            longUrl: file ? buildFileLongUrl(shortUrl) : longUrl,
+            shortUrl,
+            isFile: !!file,
+          },
+          { transaction: t },
+        )
+        if (file && !Array.isArray(file)) {
```

Comment:
why do we need to check for an array here?

Reply:
if so, should the check be performed at the start of the middleware with potential 4xx response, instead of here?